### ISSUE TITLE
Eliminate structs that wrap user-defined functions

### ISSUE THREAD

**Issue by bgoodri** [stan-dev/stan#2703]
#### Summary:

The parser should not generate structs that call user-defined functions anymore. We should use generalized lambda functions instead.

#### Description:

A user-defined function `foo` will generate additional C++ that looks something like
```
struct foo_functor__ {
    templates
    return_type
    operator()(..., std::ostream* pstream__) const {
        return foo(..., pstream__);
    }
};
```
irrespective of whether `foo_functor__` actually gets called by one of the functionals. The parser should not generate `foo_functor__`, and if functionals are used in the Stan program, then `stanc` should utilize a generalized lambda function that calls `foo`.

This would also allow us to avoid the unnecessary restriction that you cannot define multiple user-defined functions with the same name (but different signatures). See

https://github.com/stan-dev/stan/issues/1547

#### Reproducible Steps:

Parse
```
functions {
  void foo() {
  
  }
}
```

and look at the C++ code

#### Current Output:

```
struct foo_functor__ {
            void
    operator()(std::ostream* pstream__) const {
        return foo(pstream__);
    }
};
```

#### Expected Output:

Not that part

#### Additional Information:

None

#### Current Version:
v2.18.0

**Comment by bob-carpenter** [comment]
I agree this is a good idea---the lambda will essentially create the same trivial closure as we were defining manually.  The place these get used is as arguments to higher-order functions like the ODE solvers.

**Comment by VMatthijs** [comment]
@seantalts , thought this might interest you!
